File: luci-app-cloudflarespeedtest/luasrc/model/cbi/cloudflarespeedtest/third-party.lua
require("luci.sys")

local uci = luci.model.uci.cursor()

m = Map('cloudflarespeedtest', translate("Third Party Application Settings"))

s = m:section(TypedSection, "servers")
s.addremove = false
s.anonymous = true

-- [[ 第三方应用设置 ]]--

if nixio.fs.access("/etc/config/shadowsocksr") then
	s:tab("ssr", translate("ShadowSocksR Plus+"))

	o=s:taboption("ssr", Flag, "ssr_enabled",translate("ShadowSocksR Plus+ Enabled"))
	o.rmempty=true

	local ssr_server_table = {}
	uci:foreach("shadowsocksr", "servers", function(s)
		if s.alias then
			ssr_server_table[s[".name"]] = "[%s]:%s" % {string.upper(s.v2ray_protocol or s.type), s.alias}
		elseif s.server and s.server_port then
			ssr_server_table[s[".name"]] = "[%s]:%s:%s" % {string.upper(s.v2ray_protocol or s.type), s.server, s.server_port}
		end
	end)

	local ssr_key_table = {}
	for key, _ in pairs(ssr_server_table) do
		table.insert(ssr_key_table, key)
	end

	table.sort(ssr_key_table)

	o = s:taboption("ssr", DynamicList, "ssr_services",
			translate("ShadowSocksR Servers"),
			translate("Please select a service"))

	for _, key in pairs(ssr_key_table) do
		o:value(key, ssr_server_table[key])
	end
	o:depends("ssr_enabled", 1)
	o.forcewrite = true

end


if nixio.fs.access("/etc/config/passwall") then
	s:tab("passwalltab", translate("passwall"))

	o=s:taboption("passwalltab", Flag, "passwall_enabled",translate("Passwall Enabled"))
	o.rmempty=true

	local passwall_server_table = {}
	uci:foreach("passwall", "nodes", function(s)
		if s.remarks then
			passwall_server_table[s[".name"]] = "[%s]:%s" % {string.upper(s.protocol or s.type), s.remarks}
		end
	end)

	local passwall_key_table = {}
	for key, _ in pairs(passwall_server_table) do
		table.insert(passwall_key_table, key)
	end

	table.sort(passwall_key_table)

	o = s:taboption("passwalltab", DynamicList, "passwall_services",
			translate("Passwall Servers"),
			translate("Please select a service"))

	for _, key in pairs(passwall_key_table) do
		o:value(key, passwall_server_table[key])
	end
	o:depends("passwall_enabled", 1)
	o.forcewrite = true

end

if nixio.fs.access("/etc/config/passwall2") then
	s:tab("passwall2tab", translate("passwall2"))

	o=s:taboption("passwall2tab", Flag, "passwall2_enabled",translate("PassWall2 Enabled"))
	o.rmempty=true

	local passwall2_server_table = {}
	uci:foreach("passwall2", "nodes", function(s)
		if s.remarks then
			passwall2_server_table[s[".name"]] = "[%s]:%s" % {string.upper(s.protocol or s.type), s.remarks}
		end
	end)

	local passwall2_key_table = {}
	for key, _ in pairs(passwall2_server_table) do
		table.insert(passwall2_key_table, key)
	end

	table.sort(passwall2_key_table)

	o = s:taboption("passwall2tab", DynamicList, "passwall2_services",
			translate("Passwall2 Servers"),
			translate("Please select a service"))

	for _, key in pairs(passwall2_key_table) do
		o:value(key, passwall2_server_table[key])
	end
	o:depends("passwall2_enabled", 1)
	o.forcewrite = true

end

if nixio.fs.access("/etc/config/bypass") then
	s:tab("bypasstab", translate("Bypass"))

	o=s:taboption("bypasstab", Flag, "bypass_enabled",translate("Bypass Enabled"))
	o.rmempty=true

	local bypass_server_table = {}
	uci:foreach("bypass", "servers", function(s)
		if s.alias then
			bypass_server_table[s[".name"]] = "[%s]:%s" % {string.upper(s.protocol or s.type), s.alias}
		elseif s.server and s.server_port then
			bypass_server_table[s[".name"]] = "[%s]:%s:%s" % {string.upper(s.protocol or s.type), s.server, s.server_port}
		end
	end)

	local bypass_key_table = {}
	for key, _ in pairs(bypass_server_table) do
		table.insert(bypass_key_table, key)
	end

	table.sort(bypass_key_table)

	o = s:taboption("bypasstab", DynamicList, "bypass_services",
			translate("Bypass Servers"),
			translate("Please select a service"))

	for _, key in pairs(bypass_key_table) do
		o:value(key, bypass_server_table[key])
	end
	o:depends("bypass_enabled", 1)
	o.forcewrite = true

end

if nixio.fs.access("/etc/config/vssr") then
	s:tab("vssrtab", translate("Vssr"))

	o=s:taboption("vssrtab", Flag, "vssr_enabled",translate("Vssr Enabled"))
	o.rmempty=true

	local vssr_server_table = {}
	uci:foreach("vssr", "servers", function(s)
		if s.alias then
			vssr_server_table[s[".name"]] = "[%s]:%s" % {string.upper(s.protocol or s.type), s.alias}
		elseif s.server and s.server_port then
			vssr_server_table[s[".name"]] = "[%s]:%s:%s" % {string.upper(s.protocol or s.type), s.server, s.server_port}
		end
	end)

	local vssr_key_table = {}
	for key, _ in pairs(vssr_server_table) do
		table.insert(vssr_key_table, key)
	end

	table.sort(vssr_key_table)

	o = s:taboption("vssrtab", DynamicList, "vssr_services",
			translate("Vssr Servers"),
			translate("Please select a service"))

	for _, key in pairs(vssr_key_table) do
		o:value(key, vssr_server_table[key])
	end
	o:depends("vssr_enabled", 1)
	o.forcewrite = true

end

s:tab("dnshost", translate("HOST"))
o=s:taboption("dnshost", Flag, "HOST_enabled",translate("HOST Enabled"))
o=s:taboption("dnshost", Value,"host_domain",translate("Domain"))
o.rmempty=true
o:depends("HOST_enabled", 1)

s:tab("dnstab", translate("DNS"))

o=s:taboption("dnstab", Flag, "DNS_enabled",translate("DNS Enabled"))

o=s:taboption("dnstab", ListValue, "DNS_type", translate("DNS Type"))
o:value("aliyun", translate("Alibaba Cloud DNS"))
o:depends("DNS_enabled", 1)

o=s:taboption("dnstab", Value,"app_key",translate("Access Key ID"))
o.rmempty=true
o:depends("DNS_enabled", 1)
o=s:taboption("dnstab", Value,"app_secret",translate("Access Key Secret"))
o.rmempty=true
o:depends("DNS_enabled", 1)

o=s:taboption("dnstab", Value,"main_domain",translate("Main Domain"),translate("For example: test.github.com -> github.com"))
o.rmempty=true
o:depends("DNS_enabled", 1)
o=s:taboption("dnstab", DynamicList,"sub_domain",translate("Sub Domain"),translate("For example: test.github.com -> test"))
o.rmempty=true
o:depends("DNS_enabled", 1)

o=s:taboption("dnstab", ListValue, "line", translate("Lines"))
o:value("default", translate("default"))
o:value("telecom", translate("telecom"))
o:value("unicom", translate("unicom"))
o:value("mobile", translate("mobile"))
o:depends("DNS_enabled", 1)
o.default ="telecom"

s:tab("mosdns", translate("MosDNS"))
o=s:taboption("mosdns", Flag, "MosDNS_enabled",translate("MosDNS Enabled"))
o=s:taboption("mosdns", Value, "MosDNS_ip_count",translate("MosDNS IP Count"))
o.datatype ="uinteger"
o.default = 1
o:depends("MosDNS_enabled", 1)
o=s:taboption("mosdns", Flag, "openclash_restart",translate("OpenClash Restart"))
o:depends("MosDNS_enabled", 1)

return m

File: luci-app-cloudflarespeedtest/htdocs/luci-static/resources/view/cloudflarespeedtest/third-party.js
'use strict';
'require view';
'require form';
'require rpc';
'require uci';

const callListNodes = rpc.declare({
	object: 'cloudflarespeedtest',
	method: 'list_nodes',
	expect: {}
});

function addNodeValues(option, nodes) {
	nodes = nodes || [];

	for (let i = 0; i < nodes.length; i++)
		option.value(nodes[i].value, nodes[i].label);
}

return view.extend({
	load: function() {
		return Promise.all([
			uci.load('cloudflarespeedtest'),
			callListNodes()
		]);
	},

	render: function(data) {
		let m, s, o;
		const nodes = data[1] || {};

		m = new form.Map('cloudflarespeedtest', _('Third Party Application Settings'));

		s = m.section(form.NamedSection, 'servers', 'servers');
		s.addremove = false;

		if (nodes.ssr && nodes.ssr.exists) {
			s.tab('ssr', _('ShadowSocksR Plus+'));

			o = s.taboption('ssr', form.Flag, 'ssr_enabled', _('ShadowSocksR Plus+ Enabled'));
			o.rmempty = true;

			o = s.taboption('ssr', form.DynamicList, 'ssr_services',
				_('ShadowSocksR Servers'), _('Please select a service'));
			addNodeValues(o, nodes.ssr.nodes);
			o.depends('ssr_enabled', '1');
			o.forcewrite = true;
		}

		if (nodes.passwall && nodes.passwall.exists) {
			s.tab('passwalltab', _('passwall'));

			o = s.taboption('passwalltab', form.Flag, 'passwall_enabled', _('Passwall Enabled'));
			o.rmempty = true;

			o = s.taboption('passwalltab', form.DynamicList, 'passwall_services',
				_('Passwall Servers'), _('Please select a service'));
			addNodeValues(o, nodes.passwall.nodes);
			o.depends('passwall_enabled', '1');
			o.forcewrite = true;
		}

		if (nodes.passwall2 && nodes.passwall2.exists) {
			s.tab('passwall2tab', _('passwall2'));

			o = s.taboption('passwall2tab', form.Flag, 'passwall2_enabled', _('PassWall2 Enabled'));
			o.rmempty = true;

			o = s.taboption('passwall2tab', form.DynamicList, 'passwall2_services',
				_('Passwall2 Servers'), _('Please select a service'));
			addNodeValues(o, nodes.passwall2.nodes);
			o.depends('passwall2_enabled', '1');
			o.forcewrite = true;
		}

		if (nodes.bypass && nodes.bypass.exists) {
			s.tab('bypasstab', _('Bypass'));

			o = s.taboption('bypasstab', form.Flag, 'bypass_enabled', _('Bypass Enabled'));
			o.rmempty = true;

			o = s.taboption('bypasstab', form.DynamicList, 'bypass_services',
				_('Bypass Servers'), _('Please select a service'));
			addNodeValues(o, nodes.bypass.nodes);
			o.depends('bypass_enabled', '1');
			o.forcewrite = true;
		}

		if (nodes.vssr && nodes.vssr.exists) {
			s.tab('vssrtab', _('Vssr'));

			o = s.taboption('vssrtab', form.Flag, 'vssr_enabled', _('Vssr Enabled'));
			o.rmempty = true;

			o = s.taboption('vssrtab', form.DynamicList, 'vssr_services',
				_('Vssr Servers'), _('Please select a service'));
			addNodeValues(o, nodes.vssr.nodes);
			o.depends('vssr_enabled', '1');
			o.forcewrite = true;
		}

		s.tab('dnshost', _('HOST'));
		o = s.taboption('dnshost', form.Flag, 'HOST_enabled', _('HOST Enabled'));
		o = s.taboption('dnshost', form.Value, 'host_domain', _('Domain'));
		o.rmempty = true;
		o.depends('HOST_enabled', '1');

		s.tab('dnstab', _('DNS'));

		o = s.taboption('dnstab', form.Flag, 'DNS_enabled', _('DNS Enabled'));

		o = s.taboption('dnstab', form.ListValue, 'DNS_type', _('DNS Type'));
		o.value('aliyun', _('Alibaba Cloud DNS'));
		o.depends('DNS_enabled', '1');

		o = s.taboption('dnstab', form.Value, 'app_key', _('Access Key ID'));
		o.rmempty = true;
		o.depends('DNS_enabled', '1');

		o = s.taboption('dnstab', form.Value, 'app_secret', _('Access Key Secret'));
		o.rmempty = true;
		o.password = true;
		o.depends('DNS_enabled', '1');

		o = s.taboption('dnstab', form.Value, 'main_domain', _('Main Domain'),
			_('For example: test.github.com -> github.com'));
		o.rmempty = true;
		o.depends('DNS_enabled', '1');

		o = s.taboption('dnstab', form.DynamicList, 'sub_domain', _('Sub Domain'),
			_('For example: test.github.com -> test'));
		o.rmempty = true;
		o.depends('DNS_enabled', '1');

		o = s.taboption('dnstab', form.ListValue, 'line', _('Lines'));
		o.value('default', _('default'));
		o.value('telecom', _('telecom'));
		o.value('unicom', _('unicom'));
		o.value('mobile', _('mobile'));
		o.depends('DNS_enabled', '1');
		o.default = 'telecom';

		o = s.taboption('dnstab', form.Value, 'AliDNS_ip_count', _('AliDNS IP Count'));
		o.datatype = 'uinteger';
		o.default = '1';
		o.depends('DNS_enabled', '1');

		s.tab('mosdns', _('MosDNS'));
		o = s.taboption('mosdns', form.Flag, 'MosDNS_enabled', _('MosDNS Enabled'));
		o = s.taboption('mosdns', form.Value, 'MosDNS_ip_count', _('MosDNS IP Count'));
		o.datatype = 'uinteger';
		o.default = '1';
		o.depends('MosDNS_enabled', '1');
		o = s.taboption('mosdns', form.Flag, 'openclash_restart', _('OpenClash Restart'));
		o.depends('MosDNS_enabled', '1');

		s.tab('astradns', _('astra-dns'));
		o = s.taboption('astradns', form.Flag, 'AstraDNS_enabled', _('astra-dns Enabled'));
		o.rmempty = true;

		o = s.taboption('astradns', form.Value, 'AstraDNS_config', _('astra-dns Config Path'));
		o.default = '/etc/astra-dns/named.yaml';
		o.rmempty = true;
		o.depends('AstraDNS_enabled', '1');

		o = s.taboption('astradns', form.Value, 'AstraDNS_bin', _('astra-dns Binary Path'));
		o.default = '/usr/bin/astra-dns';
		o.rmempty = true;
		o.depends('AstraDNS_enabled', '1');

		return m.render();
	}
});
